compiler-core/src/javascript/typescript.rs

```diff
@@ -0,0 +1,530 @@
+            "UtfCodepoint" => {
+                self.tracker.borrow_mut().prelude_used = true;
+                "$gleam.UtfCodepoint".to_doc()
+            }
+            "String" => "string".to_doc(),
+            "Bool" => "boolean".to_doc(),
+            "BitString" => {
+                self.tracker.borrow_mut().prelude_used = true;
+                "$gleam.BitString".to_doc()
+            }
+            "List" => {
+                self.tracker.borrow_mut().prelude_used = true;
+                let arg0 = self.print(args.get(0).expect("print_prelude_type list"));
+                docvec!["$gleam.List<", arg0, ">"]
+            }
+            "Result" => {
+                self.tracker.borrow_mut().prelude_used = true;
+                let arg_ok = self.print(args.get(0).expect("print_prelude_type result ok"));
+                let arg_err = self.print(args.get(1).expect("print_prelude_type result err"));
+                docvec!["$gleam.Result<", arg_ok, ", ", arg_err, ">"]
+            }
+            // Getting here sholud mean we either forgot a built-in type or there is a
+            // compiler error
+            name => panic!("{} is not a built-in type.", name),
+        }
+    }
+
+    /// Prints a "named" programmer-defined Gleam type into the Typescript
+    /// equivalent.
+    ///
+    fn print_type_app(
+        &self,
+        name: &str,
+        args: &[Arc<Type>],
+        module: &[String],
+    ) -> Document<'static> {
+        let name = format!("{}$", ts_safe_type_name(name.to_camel_case()));
+        let name = match module == self.current_module {
+            true => Document::String(name),
+            false => {
+                // If type comes from a seperate module, use that module's nam
+                // as a Typescript namespace prefix
+                let module_name = module
+                    .last()
+                    .expect("Typescript generator could not identify imported module name.");
+                let module_name = format!("${}", module_name);
+                docvec![Document::String(module_name), ".", Document::String(name),]
+            }
+        };
+        if args.is_empty() {
+            return name;
+        }
+        // If the App type takes arguments, pass them in as Typescript generics
+        let args = concat(Itertools::intersperse(
+            args.iter().map(|a| self.print(a)),
+            ", ".to_doc(),
+        ));
+        docvec![name, "<", args, ">"]
+    }
+
+    /// Prints the Typescript type for an anonymous function (aka lambda)
+    ///
+    fn print_fn(&self, args: &[Arc<Type>], retrn: &Type) -> Document<'static> {
+        let args = concat(Itertools::intersperse(
+            args.iter()
+                .enumerate()
+                .map(|(idx, a)| docvec!["x", idx, ": ", self.print(a)]),
+            ", ".to_doc(),
+        ));
+        let retrn = self.print(retrn);
+        docvec!["(", args, ") => ", retrn]
```

Comment:
Same as with type app, it would be nice to have this wrap and indent nicely when the line is too long

Reply:
Will add a test